fish-rust/src/common.rs

```diff
@@ -1827,22 +1779,15 @@ pub fn scoped_push<Context, Accessor, T>(
     mut ctx: Context,
     accessor: Accessor,
     new_value: T,
-) -> impl Deref<Target = Context> + DerefMut<Target = Context>
+) -> impl DerefMut<Target = Context>
 where
     Accessor: Fn(&mut Context) -> &mut T,
-    T: Copy,
 {
-    let saved_value = mem::replace(accessor(&mut ctx), new_value);
-    // Store the original/root value, the function to map from the original value to the variables
-    // we are changing, and a saved snapshot of the previous values of those variables in a tuple,
-    // then use ScopeGuard's `on_drop` parameter to restore the saved values when the scope ends.
-    let scope_guard = ScopeGuard::new((ctx, accessor, saved_value), |data| {
-        let (ref mut ctx, accessor, saved_value) = data;
-        *accessor(ctx) = *saved_value;
-    });
-    // `scope_guard` would deref to the tuple we gave it, so use Projection<T> to map from the tuple
-    // `(ctx, accessor, saved_value)` to the result of `accessor(ctx)`.
-    Projection::new(scope_guard, |sg| &sg.0, |sg| &mut sg.0)
+    let saved = mem::replace(accessor(&mut ctx), new_value);
```

Comment:
just a heads up, this accessor won't work for cases where we need to acquire a lock guard or similar.
For example given a `RefCell` field `library_data`

```
    pub fn libdata_mut(&self) -> RefMut<'_, LibraryData> {
        self.library_data.borrow_mut()
    }
```

there is no kosher function that returns a mutable reference without the `RefMut`.
I worked around this by writing

```
    let parser = scoped_push_replacer(
        parser,
        |parser, new_value| std::mem::replace(&mut parser.libdata_mut().pods.is_event, new_value),
        is_event + 1,
    );
```

but I realize we should instead make it so `accessor` returns something like `impl Deref<&mut T>`